src/dmd/root/longdouble.d

```diff
@@ -186,63 +251,62 @@ nothrow @nogc pure:
 
 version(LDC)
 {
-    enum fld_eax(string arg) = "jmp L_" ~ arg ~ "+2; L_" ~ arg ~ ": mov DX, 0x28db;"; // jump to 0xdb,0x28 (fld real ptr[EAX])
-    enum fstp_eax            = "jmp L_stp+2; L_stp: mov DX, 0x38db;"; // jump to 0xdb,0x38 (fstp real ptr[EAX])
-}
-else version(D_InlineAsm_X86_64)
-{
-    enum fld_eax(string arg) = "fld real ptr [RAX];";
-    enum fstp_eax            = "fstp real ptr [RAX];";
-}
-else version(D_InlineAsm_X86)
-{
-    enum fld_eax(string arg) = "fld real ptr [EAX];";
-    enum fstp_eax            = "fstp real ptr [EAX];";
-}
+    import ldc.llvmasm;
 
-version(D_InlineAsm_X86_64)
-{
-    // longdouble_soft passed by reference
     extern(D):
     private:
-    string fld_arg(string arg)()
-    {
-        return "asm nothrow @nogc pure @trusted { mov RAX, " ~ arg ~ "; " ~ fld_eax!arg ~ " }";
-    }
-    string fstp_arg(string arg)()
-    {
-        return "asm nothrow @nogc pure @trusted { mov RAX, " ~ arg ~ "; " ~ fstp_eax ~ " }";
-    }
-    alias fld_parg = fld_arg;
-    alias fstp_parg = fstp_arg;
-    string fld_local(string arg)()
-    {
-        return "asm nothrow @nogc pure @trusted { lea RAX, " ~ arg ~ "; " ~ fld_eax!arg ~ " }";
-    }
+    string fld_arg  (string arg)() { return `__asm("fldt $0",  "*m,~{st}", &` ~ arg ~ `);`; }
+    string fstp_arg (string arg)() { return `__asm("fstpt $0", "*m,~{st}", &` ~ arg ~ `);`; }
+    string fld_parg (string arg)() { return `__asm("fldt $0",  "*m,~{st}",  ` ~ arg ~ `);`; }
+    string fstp_parg(string arg)() { return `__asm("fstpt $0", "*m,~{st}",  ` ~ arg ~ `);`; }
```

Comment:
This is much better than my dirty hack, but I couldn't get it to work when I tried. If we upgrade LDC to 1.9.0, we could also use saner dmd-like asm.

Reply:
With LDC 1.9 and fixed `extended ptr` semantics for DMD `asm` (DMD doesn't support `extended` apparently), the semantics of variables in `asm` apparently still differ. The current asm still doesn't work, I tried that first.

LDC translates these DMD `asm` fragments internally to LLVM `__asm` fragments and passes in variables as LLVM-IR-values, way before register allocation etc., so getting that right is probably not trivial.

`__asm` directly on the other hand is simpler, as you usually explicitly pass in the address of variables to be read from or written to.